File: common/src/main/java/org/anvilpowered/catalyst/common/plugin/CatalystPluginMessages.java
/*
 *     Copyright (C) 2020 STG_Allen
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.catalyst.common.plugin;

import com.google.inject.Inject;
import org.anvilpowered.anvil.api.plugin.PluginInfo;
import org.anvilpowered.anvil.api.util.TextService;
import org.anvilpowered.anvil.api.util.TimeFormatService;
import org.anvilpowered.catalyst.api.plugin.PluginMessages;

import java.time.Duration;
import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;

public class CatalystPluginMessages<TString, TCommandSource> implements PluginMessages<TString> {

    @Inject
    protected PluginInfo<TString> pluginInfo;

    @Inject
    protected TextService<TString, TCommandSource> textService;

    @Inject
    protected TimeFormatService timeFormatService;

    @Override
    public TString getBroadcast(TString message) {
        return textService.builder()
            .green().append("[Broadcast] ", message)
            .build();
    }

    @Override
    public TString getBroadcast(String message) {
        return getBroadcast(textService.deserialize(message));
    }

    @Override
    public TString getNotEnoughArgs() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("Not enough arguments!")
            .build();
    }

    @Override
    public TString getNoPermission() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append(("Insufficient permissions!"))
            .build();
    }

    @Override
    public TString getNoNickColorPermission() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("You do not have permission for a colored nickname!")
            .build();
    }

    @Override
    public TString getNoNickMagicPermission() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("You do not have permission for a magical nickname!")
            .build();
    }

    @Override
    public TString getCurrentServer(String userName, String serverName) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .gold().append(userName)
            .gray().append(" is connected to ")
            .green().append(serverName, ".")
            .build();
    }

    @Override
    public TString getMuted() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("You are muted!")
            .build();
    }

    @Override
    public TString getMuteExempt() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("This user is exempt from being muted.")
            .build();
    }

    @Override
    public TString getBanExempt() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("This user is exempt from being banned.")
            .build();
    }

    @Override
    public TString getKickExempt() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("This user is exempt from being kicked.")
            .build();
    }

    @Override
    public TString getSocialSpy(boolean enabled) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .green().append("Socialspy ", enabled ? "enabled" : "disabled")
            .build();
    }

    @Override
    public TString getStaffChat(boolean enabled) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .green().append("Staff chat ", enabled ? "enabled" : "disabled")
            .build();
    }

    @Override
    public TString getStaffChatMessageFormatted(String userName, TString message) {
        return textService.builder()
            .aqua().append("[STAFF] ")
            .light_purple().append(userName, ": ", message)
            .build();
    }

    @Override
    public TString getStaffChatMessageFormattedConsole(TString message) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .aqua().append("[STAFF] ")
            .light_purple().append("CONSOLE: ", message)
            .build();
    }

    @Override
    public TString getTeleportRequestSent(String targetUserName) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .green().append("Requested to teleport to ")
            .gold().append(targetUserName, ".")
            .build();
    }

    @Override
    public TString getTeleportRequestReceived(String requesterUserName) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .gold().append(requesterUserName)
            .green().append(" has requested to teleport to you")
            .onHoverShowText(textService.builder().green().append("Click to accept"))
            .onClickRunCommand("/tpaccept")
            .build();
    }

    @Override
    public TString getTeleportToSelf() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("You cannot teleport to yourself!")
            .build();
    }

    @Override
    public TString getTeleportRequestPending(String targetUserName) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .yellow().append("You have a pending request to " + targetUserName)
            .build();
    }

    @Override
    public TString getTeleportRequestAccepted(String sourceUserNamae) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .green().append(sourceUserNamae + " has accepted your teleportation request.")
            .build();
    }

    @Override
    public TString getSourceAcceptedTeleport(String targetUserName) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .green().append("Accepted " + targetUserName + "'s teleport request.")
            .build();
    }

    @Override
    public TString getNoPendingRequests() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .yellow().append("You don't have any pending teleportation requests!")
            .build();
    }

    @Override
    public TString getIncompatibleServerVersion() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("The server you are attempting to connect to is running a different Minecraft version!")
            .build();
    }

    private TString getForList(boolean a, boolean b, boolean c, String value) {
        String _a = a ? "swear" : "exception";
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("The ", _a, " ")
            .yellow().append(value)
            .red().append(" ", b ? c ? "is already in" : "is not in" : c ? "was added to" : "was removed from", " the ", _a, " list.")
            .build();
    }

    @Override
    public TString getExistingSwear(String swear) {
        return getForList(true, true, true, swear);
    }

    @Override
    public TString getExistingException(String exception) {
        return getForList(false, true, true, exception);
    }

    @Override
    public TString getMissingSwear(String swear) {
        return getForList(true, true, false, swear);
    }

    @Override
    public TString getMissingException(String exception) {
        return getForList(false, true, false, exception);
    }

    @Override
    public TString getNewSwear(String swear) {
        return getForList(true, false, true, swear);
    }

    @Override
    public TString getNewException(String exception) {
        return getForList(false, false, true, exception);
    }

    @Override
    public TString getRemoveSwear(String swear) {
        return getForList(true, false, false, swear);
    }

    @Override
    public TString getRemoveException(String exception) {
        return getForList(false, false, false, exception);
    }

    public String removeColor(String in) {
        return in.replace("&0", "").replaceAll("&1", "").replaceAll("&2", "")
            .replaceAll("&3", "").replaceAll("&4", "").replaceAll("&5", "")
            .replaceAll("&6", "").replaceAll("&7", "").replaceAll("&8", "")
            .replaceAll("&a", "").replaceAll("&b", "").replaceAll("&c", "")
            .replaceAll("&d", "").replaceAll("&e", "").replaceAll("&f", "")
            .replaceAll("&k", "").replaceAll("&l", "").replaceAll("&m", "")
            .replaceAll("&n", "").replaceAll("&o", "").replaceAll("&r", "")
            .replaceAll("&k", "").replaceAll("&9", "");
    }

    @Override
    public TString banCommandUsage() {
        return textService.deserialize("&4Usage:  &e/ban <user> [reason]");
    }

    @Override
    public TString tempBanCommandUsage() {
        return textService.deserialize("&4Usage:  &e/tempban <user> <duration> [reason]");
    }

    @Override
    public TString unbanCommandUsage() {
        return textService.deserialize("&4Usage: &e/unban <user>");
    }

    @Override
    public TString muteCommandUsage() {
        return textService.deserialize("&4Usage: &e/mute <user> [reason]");
    }

    @Override
    public TString tempMuteCommandUsage() {
        return textService.deserialize("&4Usage: &e/tempmute <user> <duration> [reason]");
    }

    @Override
    public TString unMuteCommandUsage() {
        return textService.deserialize("&4Usage: &e/unmute <player>");
    }

    @Override
    public TString kickCommandUsage() {
        return textService.deserialize("&4Usage: &e/kick <user> [reason]");
    }

    @Override
    public TString findCommandUsage() {
        return textService.deserialize("&4Usage: /find <user>");
    }

    @Override
    public TString sendCommandUsage() {
        return textService.deserialize("&4Usage: /send <user> <server>");
    }

    @Override
    public TString messageCommandUsage() {
        return textService.deserialize("&4Usage: /message <user> <message>");
    }

    @Override
    public TString nickNameCommandUsage() {
        return textService.deserialize("&4Usage: /nick <nickname>");
    }

    @Override
    public TString broadcastCommandUsage() {
        return textService.deserialize("&4Usage: /broadcast <message>");
    }

    @Override
    public TString infoCommandUsage() {
        return textService.deserialize("&4Usage: /info <user>");
    }

    @Override
    public TString swearAddCommandUsage() {
        return textService.deserialize("&4Usage: /swear (add|remove <word>) |list");
    }

    @Override
    public TString exceptionAddCommandUsage() {
        return textService.deserialize("&4Usage: /exception (add|remove <word>) |list");
    }

    @Override
    public TString ignoreCommandUsage() {
        return textService.deserialize("&4Usage: /ignore (playername)");
    }

    @Override
    public TString ignoreExempt() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("This user is exempt from being ignored.")
            .build();
    }

    @Override
    public TString offlineOrInvalidPlayer() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .yellow().append("Invalid or offline player!")
            .build();
    }

    @Override
    public TString getBanMessage(String reason, Instant endUtc) {
        return textService.builder()
            .red().append("You have been banned for: ", textService.deserialize(reason))
            .yellow().append("\n\nFor another ", timeFormatService.format(Duration.between(OffsetDateTime.now(ZoneOffset.UTC).toInstant(), endUtc)))
            .append("\n\nUntil ", timeFormatService.format(endUtc))
            .build();
    }

    @Override
    public TString getMuteMessage(String reason, Instant endUtc) {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .red().append("You have been muted for: ", textService.deserialize(reason))
            .yellow().append("\nFor another ", timeFormatService.format(Duration.between(OffsetDateTime.now(ZoneOffset.UTC).toInstant(), endUtc)))
            .build();
    }

    @Override
    public TString getInvalidServer() {
        return textService.builder()
            .append(pluginInfo.getPrefix())
            .yellow().append("Invalid server!")
            .build();
    }
}

File: common/src/main/kotlin/org/anvilpowered/catalyst/common/plugin/CatalystPluginMessages.kt
/*
 * Catalyst - AnvilPowered
 *   Copyright (C) 2020
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see https://www.gnu.org/licenses/.
 */

package org.anvilpowered.catalyst.common.plugin

import com.google.inject.Inject
import org.anvilpowered.anvil.api.util.TextService
import org.anvilpowered.anvil.api.util.TimeFormatService
import org.anvilpowered.catalyst.api.plugin.PluginMessages
import java.time.Duration
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset

class CatalystPluginMessages<TString, TCommandSource> @Inject constructor(
  private val textService: TextService<TString, TCommandSource>,
  private val timeFormatService: TimeFormatService
) : PluginMessages<TString> {

  override fun getBroadcast(message: TString): TString {
    return textService.builder()
      .green().append("[Broadcast] ", message)
      .build()
  }

  override fun getBroadcast(message: String): TString = getBroadcast(textService.deserialize(message))

  override fun getNotEnoughArgs(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("Not enough arguments!")
      .build()
  }

  override fun getNoPermission(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("Insufficient permissions!")
      .build()
  }

  override fun getNoServerPermission(serverName: String): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("You do not have permission to enter ")
      .gold().append(serverName)
      .red().append("!")
      .build()
  }

  override fun getNoNickColorPermission(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("You do not have permission for a colored nickname!")
      .build()
  }

  override fun getNoNickMagicPermission(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("You do not have permission for a magical nickname!")
      .build()
  }

  override fun getCurrentServer(userName: String, serverName: String): TString {
    return textService.builder()
      .appendPrefix()
      .gold().append(userName)
      .green().append(" is connected to ")
      .gold().append(serverName, ".")
      .build()
  }

  override fun getMuted(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("You are muted!")
      .build()
  }

  override fun getMuteExempt(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("This user is exempt from being muted.")
      .build()
  }

  override fun getBanExempt(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("This user is exempt from being banned.")
      .build()
  }

  override fun getKickExempt(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("This user is exempt from being kicked.")
      .build()
  }

  override fun getSocialSpy(enabled: Boolean): TString {
    val builder = textService.builder()
      .appendPrefix()
      .yellow().append("SocialSpy ")
    return if (enabled) {
      builder.green().append("enabled").build()
    } else builder.red().append("disabled").build()
  }

  override fun getStaffChat(enabled: Boolean): TString {
    val builder = textService.builder()
      .appendPrefix()
      .yellow().append("Staff Chat ")
    return if (enabled) {
      builder.green().append("enabled").build()
    } else builder.red().append("disabled").build()
  }

  override fun getStaffChatMessageFormatted(userName: String, message: TString): TString {
    return textService.builder()
      .aqua().append("[STAFF] ")
      .light_purple().append(userName, ": ", message)
      .build()
  }

  override fun getStaffChatMessageFormattedConsole(message: TString): TString {
    return textService.builder()
      .appendPrefix()
      .aqua().append("[STAFF] ")
      .light_purple().append("CONSOLE: ", message)
      .build()
  }

  override fun getTeleportRequestSent(targetUserName: String): TString {
    return textService.builder()
      .appendPrefix()
      .green().append("Requested to teleport to ")
      .gold().append(targetUserName, ".")
      .build()
  }

  override fun getTeleportRequestReceived(requesterUserName: String): TString {
    return textService.builder()
      .appendPrefix()
      .gold().append(requesterUserName)
      .green().append(" has requested to teleport to you")
      .onHoverShowText(textService.builder().green().append("Click to accept"))
      .onClickRunCommand("/tpaccept")
      .build()
  }

  override fun getTeleportToSelf(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("You cannot teleport to yourself!")
      .build()
  }

  override fun getIncompatibleServerVersion(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("The server you are attempting to connect to is running a different Minecraft version!")
      .build()
  }

  private fun getForList(a: Boolean, b: Boolean, c: Boolean, value: String): TString {
    val _a = if (a) "swear" else "exception"
    return textService.builder()
      .appendPrefix()
      .red().append("The ", _a, " ")
      .yellow().append(value)
      .red().append(" ", if (b) if (c) "is already in" else "is not in" else if (c) "was added to" else "was removed from", " the ", _a, " list.")
      .build()
  }

  override fun getExistingSwear(swear: String): TString = getForList(true, true, true, swear)
  override fun getExistingException(exception: String): TString = getForList(false, true, true, exception)
  override fun getMissingSwear(swear: String): TString = getForList(true, true, false, swear)
  override fun getMissingException(exception: String): TString = getForList(false, true, false, exception)
  override fun getNewSwear(swear: String): TString = getForList(true, false, true, swear)
  override fun getNewException(exception: String): TString = getForList(false, false, true, exception)
  override fun getRemoveSwear(swear: String): TString = getForList(true, false, false, swear)
  override fun getRemoveException(exception: String): TString = getForList(false, false, false, exception)

  override fun banCommandUsage(): TString {
    return textService.builder()
      .red().append(textService.of("Usage:"))
      .yellow().append("/ban <user> [reason]")
      .build()
  }

  override fun tempBanCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/tempban <user> <duration> [reason]")
      .build()
  }

  override fun unbanCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/unban <user>")
      .build()
  }

  override fun muteCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/mute <user> [reason]")
      .build()
  }

  override fun tempMuteCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/tempmute <user> <duration> [reason]")
      .build()
  }

  override fun unMuteCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/unmute <player>")
      .build()
  }

  override fun kickCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/kick <user> [reason]")
      .build()
  }

  override fun findCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/find <user>")
      .build()
  }

  override fun sendCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/send <user> <server>")
      .build()
  }

  override fun messageCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/message <user> <message>")
      .build()
  }

  override fun replyCommandUsage(): TString {
    return textService.deserialize("&4Usage: /reply <message>")
  }

  override fun nickNameCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/nick <nickname>")
      .build()
  }

  override fun deleteNickOtherCommandUsage(): TString {
    return textService.deserialize("&4Usage: /delnick other <player>")
  }

  override fun broadcastCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/broadcast <message>")
      .build()
  }

  override fun infoCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/info <user>")
      .build()
  }

  override fun channelEditStartUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/channel edit start <channel>")
      .build()
  }

  override fun channelEditPropertyUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/channel edit property <name> <\"value\">")
      .build()
  }

  override fun swearAddCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/swear (add|remove <word>) | list")
      .build()
  }

  override fun exceptionAddCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/exception (add|remove <word) | list")
      .build()
  }

  override fun ignoreCommandUsage(): TString {
    return textService.builder()
      .red().append("Usage: ")
      .yellow().append("/ignore <player>")
      .build()
  }

  override fun ignoreExempt(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("This user is exempt from being ignored.")
      .build()
  }

  override fun offlineOrInvalidPlayer(): TString {
    return textService.builder()
      .appendPrefix()
      .yellow().append("Invalid or offline player!")
      .build()
  }

  override fun messageSelf(): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("You cannot send private messages to yourself!")
      .build()
  }

  override fun getBanMessage(reason: String, endUtc: Instant): TString {
    return textService.builder()
      .red().append("You have been banned for: ", textService.deserialize(reason))
      .yellow().append(
        "\n\nFor another ",
        timeFormatService.format(Duration.between(OffsetDateTime.now(ZoneOffset.UTC).toInstant(), endUtc)).withoutNano()
      )
      .append("\n\nUntil ", timeFormatService.format(endUtc).withoutNano())
      .build()
  }

  override fun getMuteMessage(reason: String, endUtc: Instant): TString {
    return textService.builder()
      .appendPrefix()
      .red().append("You have been muted for: ", textService.deserialize(reason))
      .yellow()
      .append("\nFor another ", timeFormatService.format(Duration.between(OffsetDateTime.now(ZoneOffset.UTC).toInstant(), endUtc)).withoutNano())
      .build()
  }

  override fun getInvalidServer(): TString {
    return textService.builder()
      .appendPrefix()
      .yellow().append("Invalid server!")
      .build()
  }
}

